src/NSrtm.Core/Pgm/DataDesciption/PgmDataDescriptionExtractor.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Text.RegularExpressions;
+
+namespace NSrtm.Core.Pgm
+{
+    internal static class PgmPreamble
+    {
+        public static readonly IReadOnlyDictionary<string, Regex> Extractor =
+            new Dictionary<string, Regex>
+            {
+                {"MagicNumber", new Regex(@"^(P5)[^$]", RegexOptions.Multiline)},
+                {"Offset", new Regex(@"^#[\s]Offset[\s]+(?<offset>.+)$", RegexOptions.Multiline)},
+                {"Scale", new Regex(@"^#[\s]Scale[\s]+(?<scale>.+)$", RegexOptions.Multiline)},
+                {"Origin", new Regex(@"^#[\s]Origin[\s](?<lat>\d+)\D+(?<lon>\d+)", RegexOptions.Multiline)},
+                {"PointParameters", new Regex(@"(\n|\r|\r\n)(?<width>\d+)\s+(?<height>\d+)(\n|\r|\r\n)(?<maxValue>\d+)", RegexOptions.Multiline)},
+            };
+    }
+
+    internal static class PgmDataDescriptionExtractor
+    {
+        public static PgmDataDescription FromStream(Stream stream)
+        {
+            var preamble = extractPreambleFromStream(stream);
+            return getConstatantsFromPreamble(preamble);
+        }
+
+        private static string extractPreambleFromStream(Stream stream)
+        {
+            using (var reader = new BinaryReader(stream, Encoding.UTF8, true))
+            {
+                var preamble = new StringBuilder(reader.ReadLine());
+                string lastLine;
+                do
+                {
+                    lastLine = reader.ReadLine();
+                    preamble.Append(lastLine + "\n");
+                } while (lastLine.StartsWith("#", StringComparison.CurrentCulture));
+                preamble.Append(reader.ReadLine());
+                return preamble.ToString();
+            }
+        }
+
+        internal static PgmDataDescription getConstatantsFromPreamble(string preamble)
+        {
+            var matches = PgmPreamble.Extractor.Select(kp => kp.Value.Match(preamble))
+                                     .Zip(PgmPreamble.Extractor.Keys, (v, k) => new {k, v})
+                                     .ToDictionary(x => x.k, x => x.v);
+            foreach (var match in matches.Where(match => !match.Value.Success))
+            {
+                throw new ArgumentException(String.Format("Can not extract value {0} from the preamble", match.Key));
+            }
+
+            var offset = Convert.ToDouble(matches["Offset"].Groups["offset"].Value, CultureInfo.InvariantCulture);
```

Comment:
Please use typename.Parse (`double.Parse). 

For reference:
http://stackoverflow.com/questions/586436/double-tryparse-or-convert-todouble-which-is-faster-and-safer